allennlp/models/semantic_parsing/text2sql_parser.py

```diff
@@ -0,0 +1,399 @@
+import torch
+
+from allennlp.data import Vocabulary
+from allennlp.data.fields.production_rule_field import ProductionRule
+from allennlp.semparse.executors import SqlExecutor
+from allennlp.models.model import Model
+from allennlp.modules import Attention, Seq2SeqEncoder, TextFieldEmbedder, Embedding
+from allennlp.nn import util
+from allennlp.nn.initializers import InitializerApplicator
+from allennlp.nn.regularizers import RegularizerApplicator
+from allennlp.semparse.contexts.sql_context_utils import action_sequence_to_sql
+from allennlp.state_machines.states import GrammarBasedState
+from allennlp.state_machines.transition_functions import BasicTransitionFunction
+from allennlp.state_machines import BeamSearch
+from allennlp.state_machines.trainers import MaximumMarginalLikelihood
+from allennlp.state_machines.states import GrammarStatelet, RnnStatelet
+from allennlp.training.metrics import Average
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+@Model.register("text2sql_parser")
+class Text2SqlParser(Model):
+    """
+    Parameters
+    ----------
+    vocab : ``Vocabulary``
+    utterance_embedder : ``TextFieldEmbedder``
+        Embedder for utterances.
+    action_embedding_dim : ``int``
+        Dimension to use for action embeddings.
+    encoder : ``Seq2SeqEncoder``
+        The encoder to use for the input utterance.
+    decoder_beam_search : ``BeamSearch``
+        Beam search used to retrieve best sequences after training.
+    max_decoding_steps : ``int``
+        When we're decoding with a beam search, what's the maximum number of steps we should take?
+        This only applies at evaluation time, not during training.
+    input_attention: ``Attention``
+        We compute an attention over the input utterance at each step of the decoder, using the
+        decoder hidden state as the query.  Passed to the transition function.
+    database_file: ``str``, required.
+        The path of the SQLite database when evaluating SQL queries. SQLite is disk based, so we need
+        the file location to connect to it.
+    add_action_bias : ``bool``, optional (default=True)
+        If ``True``, we will learn a bias weight for each action that gets used when predicting
+        that action, in addition to its embedding.
+    dropout : ``float``, optional (default=0)
+        If greater than 0, we will apply dropout with this probability after all encoders (pytorch
+        LSTMs do not apply dropout to their last layer).
+    """
+    def __init__(self,
+                 vocab: Vocabulary,
+                 utterance_embedder: TextFieldEmbedder,
+                 action_embedding_dim: int,
+                 encoder: Seq2SeqEncoder,
+                 decoder_beam_search: BeamSearch,
+                 max_decoding_steps: int,
+                 input_attention: Attention,
+                 database_file: str,
+                 add_action_bias: bool = True,
+                 training_beam_size: int = None,
```

Comment:
This isn't needed, is it?  This is only for when you have more than one logical form at training time.

Reply:
oh yep good point, thanks.

Reply:
One quick point: Kevin found that passing `None` to `MaximumMarginalLikelihood` gave a huge slowdown for some reason I don't understand yet.  You might still want to pass `1`.  It just doesn't need to be a constructor parameter.